### ISSUE TITLE
Make less/no assumption about the data

### ISSUE THREAD

**Issue by wiseodd** [f-dangel/curvlinops#86]
Currently, `curvlinops` assumes that `X` and `y` are both `torch.Tensor`. This works for the old deep learning paradigm. But with the rise of LLMs and other complicated models, one should make that assumption. 

For instance, the input of a Huggingface model is a `UserDict`:
```python
data = UserDict({
    'input_ids': torch.LongTensor(...),
    'attention_mask': torch.LongTensor(...),
    'labels': torch.LongTensor(...)
})
```
In this case, one can extract `X` and `y` via (example from [`laplace-torch`](https://github.com/aleximmer/Laplace/blob/2f1860bfeadbc933e91347a694be8ab7dda2e1f7/laplace/baselaplace.py#L465-L467))
```python
if isinstance(data, UserDict)  or isinstance(data, dict): # To support Huggingface dataset
    X, y = data, data['labels'].to(self._device)
```

However, `curvlinops` is strongly assuming `X` to be a tensor, see for example [here](https://github.com/f-dangel/curvlinops/blob/5852711aedf2728bc609fabfa95eac00da1beb63/curvlinops/_base.py#L330-L332) and [here](https://github.com/f-dangel/curvlinops/blob/5852711aedf2728bc609fabfa95eac00da1beb63/curvlinops/_base.py#L107-L111) and [here](https://github.com/f-dangel/curvlinops/blob/5852711aedf2728bc609fabfa95eac00da1beb63/curvlinops/_base.py#L371-L373).

I think the best way to circumvent this is to not assume anything about `X` (preferably also `y`, e.g. for multi-output models).

**Comment by runame** [comment]
I agree that for many use cases the assumptions about the data are too strong. Properly addressing this requires significant changes to `curvlinops`, but is probably worth it.

I currently use a hacky workaround where I wrap the data loader and the individual batches with custom classes, such that each batch will follow the required `X, y` format, where `X` supports `.to()`, `.device`, `.dtype` etc. This also requires changing the model to deal with these wrapped inputs.

**Comment by f-dangel** [comment]
I would currently recommend writing a model and dataloader wrapper to fit them into the assumed interface `loss = loss_func(model(X), y) where (X, y) = next(data_loader)`. 

I would also like to ping @mf-silva as he has already some experience with using `curvlinops` in the context of LLMs. Short-term, we could think about documenting how to handle LLMs in a tutorial. An internal refactoring will involve a lot of changes as the assumption is made by the current base class from which almost all linear operators inherit.

**Comment by wiseodd** [comment]
As far as I see, in `curvlinops/_base.py`, the `X` being a tensor assumption is only used (i) for inferring the data shape and (ii) moving it to the correct device.

https://github.com/f-dangel/curvlinops/blob/0c8439168e457bf36917bea03100436bca6daef8/curvlinops/_base.py#L107-L111

https://github.com/f-dangel/curvlinops/blob/0c8439168e457bf36917bea03100436bca6daef8/curvlinops/_base.py#L371-L373

https://github.com/f-dangel/curvlinops/blob/0c8439168e457bf36917bea03100436bca6daef8/curvlinops/_base.py#L330-L332


For (i), we can simply remove `.to(...)` if `X` is not a tensor, and let the model handle it. Example [here](https://github.com/wiseodd/lapeft-bayesopt/blob/5a9eec65e2f075531e51ec45b5285b76774b5703/lapeft_bayesopt/foundation_models/llama2.py#L27-L32). This is a much lower friction than requiring the user to write/use a wrapper dataloader.

For (ii), wouldn't it make sense to simply require the user to input the batch shape if again `X` is not a tensor? Alternatively ask the user to provide a closure to obtain a batch shape from a non-tensor `X`. E.g. 
```python
batch_shape_huggingface: Callable[[UserDict], int] = lambda X: X['input_ids'].shape[0]
```
